### PR TITLE
Infer to partially homomorphic types (such as Pick<T, K>)

### LINKED ISSUE #29765: Change in behavior around keyof any inference
**TypeScript Version:**  3.4.0-dev.20190202

**Search Terms:** any, array, argument

**Code**

```ts
function getProps<T, K extends keyof T>(obj: T, list: K[]): Pick<T, K> {
    return {} as any;
}

const myAny: any = {};


// TS 3.2.2 - getProps returns Pick<any, "foo" | "bar">
// TS 3.4   - getProps returns Pick<any, "foo" | "bar">
const o1 = getProps(myAny, ['foo', 'bar']);

// TS 3.2.2 - getProps returns Pick<any, "foo" | "bar">
// TS 3.4   - getProps returns Pick<any, string>
   // └ 🚨 Type 'Pick<any, string>' is missing the following properties
   //   from type '{ foo: any; bar: any; }': foo, bar [2739]
const o2: { foo: any; bar: any } = getProps(myAny, ['foo', 'bar']);

```

**Expected behavior:**

I expect `o2` above to type-check successfully in 3.4

**Actual behavior:**

**Playground Link:**  NOTE: does not demonstrate the reported behavior, since it's only reproducible in recent 3.4 nightly builds

https://www.typescriptlang.org/play/index.html#src=function%20getProps%3CT%2C%20K%20extends%20keyof%20T%3E(obj%3A%20T%2C%20list%3A%20K%5B%5D)%3A%20Pick%3CT%2C%20K%3E%20%7B%0D%0A%20%20%20%20return%20%7B%7D%20as%20any%3B%0D%0A%7D%0D%0A%0D%0Aconst%20myAny%3A%20any%20%3D%20%7B%7D%3B%0D%0A%0D%0A%2F%2F%20TS%203.2.2%20-%20getProps%20returns%20Pick%3Cany%2C%20%22foo%22%20%7C%20%22bar%22%3E%0D%0A%2F%2F%20TS%203.4%20%20%20-%20getProps%20returns%20Pick%3Cany%2C%20string%3E%0D%0Aconst%20o2%3A%20%7B%20foo%3A%20any%3B%20bar%3A%20any%20%7D%20%3D%20getProps(myAny%2C%20%5B'foo'%2C%20'bar'%5D)%3B%0D%0A%0D%0A%2F%2F%20TS%203.2.2%20-%20getProps%20returns%20Pick%3Cany%2C%20%22foo%22%20%7C%20%22bar%22%3E%0D%0A%2F%2F%20TS%203.4%20%20%20-%20getProps%20returns%20Pick%3Cany%2C%20%22foo%22%20%7C%20%22bar%22%3E%0D%0Aconst%20o3%20%3D%20getProps(myAny%2C%20%5B'foo'%2C%20'bar'%5D)%3B%0D%0A

**Related Issues:** #29698 #25759

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we improve type inference for partially homomorphic types such as `Pick<T, K>`. Now, when inferring to a mapped type `{ [P in K]: X }`, where `K` is a type parameter `K extends keyof T`, we make the same inferences as we would for a homomorphic mapped type `{ [P in keyof T]: X }`.

Fixes #29765.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -14583,31 +14596,31 @@ namespace ts {
...
                     const savePriority = priority;
                     priority |= InferencePriority.MappedTypeConstraint;
                     inferFromTypes(getIndexType(source), constraintType);
                     priority = savePriority;
+                    // If K is constrained to an index type keyof T, where T is a type parameter, proceed to make
+                    // the same inferences as we would for a homomorphic mapped type { [P in keyof T]: X } (this
+                    // enables us to make meaningful inferences when the target is a Pick<T, K>). Otherwise, infer
+                    // from a union of the property types in the source to the template type X.
+                    const extendedConstraint = getConstraintOfType(constraintType);
+                    if (extendedConstraint && extendedConstraint.flags & TypeFlags.Index) {
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/checker.ts]: We don't do that for ordinary homomorphic types so I'm not sure why we'd do that here. Also, not clear how we would divide the inferences between `A` and `B`.

[the remark this answers is not in the document]
Comment (weswigham) [reply src/compiler/checker.ts]: We _do_ do that for ordinary homomorphic mapped types. It's why `constraintType` is passed into `inferFromMappedTypeConstraint`.

[the remark this answers is not in the document]
Comment (weswigham) [reply src/compiler/checker.ts]: Like, in this inference function we break down `keyof A | keyof B`, and we break down `(T extends keyof A) | (U extends keyof B)`, I don't see why we would not also break down `T extends (keyof A | keyof B)`.

[the remark this answers is not in the document]
Comment (weswigham) [reply src/compiler/checker.ts]:
```diff
-                     if (extendedConstraint && extendedConstraint.flags & TypeFlags.Index) {
-                         inferToHomomorphicMappedType(source, target, <IndexType>extendedConstraint);
-                         return true;
+                     if (extendedConstraint) {
+                         return inferFromMappedTypeConstraint(source, target, extendedConstraint);
```
maybe.

[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/checker.ts]: I see what you're getting at. We really need to repeat the whole process for the constraint.

**Pull request merged**